src/tomography.jl

```diff
@@ -0,0 +1,244 @@
+# n : number of qubits
+
+#Construct the predictor matrix.  Each row is an experiment.  The number of
+#columns is 4^n for the size of the vectorized density matrix.
+#%First transform the measurement operators by the readout pulses to create
+#the effective measurement operators and then flatten into row of the
+#predictor matrix
+@printf("Setting up predictor matrix....");
+    predictorMat = complex(zeros(length(expResults), 4^n));
+for expct = 1:length(expResults)
+
+        tmp = transpose(measPulseUs[Int(measPulseMap[Int(expct)])]'*measOps[Int(measOpMap[Int(expct)])]*measPulseUs[Int(measPulseMap[Int(expct)])]);
+        predictorMat[expct,:] = tmp[:];
+end
+@printf("Done!\n")
+    invVarMat = inv(varMat);
+    A = predictorMat'*invVarMat*predictorMat
+    B = predictorMat'*invVarMat*expResults
+    rhoLSQ = A\B
+    rhoLSQ = reshape(rhoLSQ, 2^n, 2^n);
+    return rhoLSQ
+end
+
+function QST_SDP(expResults, varMat, measPulseMap, measOpMap, measPulseUs, measOps, n)
+#Function to perform constrained SDP optimization of a physical density matrix
+#consitent with the data.
+#
+# expResults : structure array (length total number of experiments)
+#   each structure containts fields data, measPulse, measOperator
+# measPulses : cell array of unitaries of measurment pulses
+# measOps : cell array of measurment operators
+# n : number of qubits
+
+#Construct the predictor matrix.  Each row is an experiment.  The number of
+#columns is 4^n for the size of the vectorized density matrix.
+#First transform the measurement operators by the readout pulses to create
+#the effective measurement operators and then flatten into row of the
+#predictor matrix
+
+@printf("Setting up predictor matrix....");
+ predictorMat = complex(zeros(length(expResults), 4^n));for expct = 1:length(expResults)
+  tmp = transpose(measPulseUs[Int(measPulseMap[Int(expct)])]'*measOps[Int(measOpMap[Int(expct)])]*measPulseUs[Int(measPulseMap[Int(expct)])]);
+     predictorMat[expct,:] = tmp[:];
+end
+@printf("Done!\n.")
+
+@printf("SDP optimizing...")
+
+solver = SCSSolver(verbose=0, max_iters=10_000, eps = 1e-8)
+
+invVarMat = inv(varMat);
+invVarMat = real(sqrtm(real(sqrtm(invVarMat'*invVarMat))));
+
+
+ρr = Variable(2^n, 2^n)
+ρi = Variable(2^n, 2^n)
+
+constraints = trace(ρr) == 1
+constraints += trace(ρi) == 0
+constraints += isposdef([ρr ρi; -ρi ρr])
+
+#We want to minimize the difference between predicted results and experimental results
+problem = minimize( vecnorm(invVarMat*(expResults - [real(predictorMat) imag(predictorMat)]*[vec(ρr); vec(ρi)]), 2)^2, constraints )
+solve!(problem, solver)
+
+@printf("Done\n")
+return (ρr.value - 1im*ρi.value)
+end
+
+
+function analyzeStateTomo2(data, nbrQubits, nbrPulses, nbrCalRepeats)
+
+    numMeas = length(data);
+    measOps = Dict();
+    tomoData = Array(Float64[]);
+    varData = Array(Float64[]);
+    if haskey(data, "data")
+        datatemp = Dict()
+        datatemp[1]=data
```

Comment:
Again, this construction loses all type information about the elements. Also, `datatemp` is not defined if we don't take the branch.